### PR TITLE
lib/posix-futex: Clear thread's objects upon thread exit

### PR DESCRIPTION
**Pull request by michpappas**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Description of changes

If a thread creates a futex that is not cleared upon application exit that futex will remain in futex_list after the unmap of the thread's stack causing an abort on the next iteration of the list.

Clear thread's objects upon thread exit.

### CODE REVIEW

**Review by mariapana — APPROVED** [review]
Looks good.
Reviewed-by: Maria Pana <maria.pana4@gmail.com>

File: lib/posix-futex/futex.c [hunk 1]
```diff
@@ -382,14 +382,25 @@ UK_LLSYSCALL_R_DEFINE(pid_t, set_tid_address, pid_t *, tid_ref)
 	return self_tid;
 }
 
-/* Thread exit handler that clears child TID at the stored reference */
-static void pfutex_child_cleartid_term(struct uk_thread *child __unused)
+static void thread_exit_handler(struct uk_thread *child __unused)
```
Comment (mariasfiraiala) [inline lib/posix-futex/futex.c]: Is the `__unused` attribute still needed now? As far as I can see `child` is used at line 399.

**Review by mariasfiraiala — COMMENTED** [review]
All good on my side, besides a small comment.

**Review by flpostolache — COMMENTED** [review]
I think a note to mention that this function is not SMP safe should be added so that when the full SMP support is added, this function would be also changed.
From my understanding, the `list_node` will break if two threads will try to delete two adjacent futexes at the same time on two different cores.

File: lib/posix-futex/futex.c (hunk 1 above)
Comment (michpappas) [reply lib/posix-futex/futex.c]: Good catch, thanks! 👍🏼

**Comment by michpappas** [comment]
> I think a note to mention that this function is not SMP safe should be added so that when the full SMP support is added, > this function would be also changed.
> From my understanding, the list_node will break if two threads will try to delete two adjacent futexes at the same time on two different cores.

@flpostolache thanks for pointing this out. I added a calls to acquire and release `futex_llist_lock` to protect the list. 

I also added a `break` statement once the thread's entry is found, as pointed out during a private discussion with @mogasergiu, and removed `__unused` as pointed out by @mariasfiraiala.

**Review by flpostolache — APPROVED** [review]
Thanks for the changes! Looks good now.

Reviewed-by: Florin Postolache <florin.postolache.of@gmail.com>

**Review by mariasfiraiala — APPROVED** [review]
All good on my side now too, thanks!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**